src/cplscheme/SerialCouplingScheme.hpp

```diff
@@ -45,28 +47,48 @@ class SerialCouplingScheme : public BaseCouplingScheme {
       m2n::PtrM2N                   m2n,
       constants::TimesteppingMethod dtMethod,
       CouplingMode                  cplMode,
-      int                           maxIterations = 1);
+      int                           maxIterations = -1);
+
+  friend struct CplSchemeTests::SerialImplicitCouplingSchemeTests::testExtrapolateData; // For whitebox tests
+
+private:
 
   /**
    * @brief TODO
-   *
-   * @param[in] startTime TODO
-   * @param[in] startTimeWindow TODO
    */
-  virtual void initialize(double startTime, int startTimeWindow);
+  logging::Logger _log{"cplschemes::SerialCouplingSchemes"};
 
-  virtual void initializeData();
+  /**
+   * @brief TODO
+   */
+  std::pair<bool, bool> doAdvance() override;
 
-  virtual void advance();
+  DataMap &getAcceleratedData() override
+  {
+    return getSendData();
+  }
 
-  logging::Logger _log{"cplschemes::SerialCouplingSchemes"};
+  /**
+* @brief TODO
+*/
+  void initializeImplicit() override;
```

Comment:
All three implementations are similar, but the differ in some details:

## `ParallelCouplingScheme::initializeImplicit()`
```
void ParallelCouplingScheme::initializeImplicit()
{
  PRECICE_CHECK(not getSendData().empty(), "No send data configured. Use explicit scheme for one-way coupling.");
  if (not doesFirstStep()) {                 // second participant
    setupConvergenceMeasures();              // needs _couplingData configured
    mergeData();                             // merge send and receive data for all pp calls
    setupDataMatrices(getAcceleratedData()); // Reserve memory and initialize data with zero
    if (getAcceleration()) {
      getAcceleration()->initialize(getAcceleratedData()); // Reserve memory, initialize
    }
  }
}
```

## `SerialCouplingScheme::initializeImplicit()`
```
void SerialCouplingScheme::initializeImplicit()
{
  PRECICE_CHECK(not getSendData().empty(), "No send data configured! Use explicit scheme for one-way coupling.");
  if (not doesFirstStep()) {
    if (not getConvergenceMeasures().empty()) {
      setupConvergenceMeasures();              // needs _couplingData configured
      setupDataMatrices(getAcceleratedData()); // Reserve memory and initialize data with zero
    }
    if (getAcceleration()) {
      getAcceleration()->initialize(getAcceleratedData()); // Reserve memory, initialize
    }
  } else if (getAcceleration() && not getAcceleration()->getDataIDs().empty()) {
    int dataID = *(getAcceleration()->getDataIDs().begin());
    PRECICE_CHECK(getSendData(dataID) == nullptr,
                  "In case of serial coupling, acceleration can be defined for "
                      << "data of second participant only!");
  }
}
```

## `MultiCouplingScheme::initializeImplicit()`
```
void MultiCouplingScheme::initializeImplicit()
{
  mergeData();                 // merge send and receive data for all pp calls
  setupConvergenceMeasures();  // needs _couplingData configured
  setupDataMatrices(_allData); // Reserve memory and initialize data with zero
  if (getAcceleration().get() != nullptr) {
    PRECICE_CHECK(getAcceleration()->getDataIDs().size() >= 3,
                  "For parallel coupling, the number of coupling data vectors has to be at least 3, not: "
                      << getAcceleration()->getDataIDs().size());
    getAcceleration()->initialize(_allData); // Reserve memory, initialize
  }
}
```

I will post the steps for moving the implementation of `initializeImplicit` into base class in this conversation.

Reply:
`mergeData` works differently for `ParallelCouplingScheme` and `MultiCouplingScheme`. It does not even exist for `SerialCouplingScheme`. I tried to improve this situation via 310a48a44a449b507d8b6dfc9791602a77ea7323, even though introducing `mergeData()` in `SerialCouplingScheme` for the sake of following the class hierarchy looks not ideal.

Reply:
f1b8551ab24af0c16d1faf2f28e12770e837df02 homogenizes the three different lead implementations.

Reply:
a6b87aaa48bc538321bb47b554f38bd5acca7d28 puts all the actual functionality into `BaseCouplingScheme`. Only some checks are left in the leaf classes (`checkInitializationAcceleration`). @uekerman is there a better place to perform these checks? Maybe the constructors? Then we could entirely get rid of the function.

Reply:
I had to revert a6b87aa via f3bd48c, since the check `PRECICE_CHECK(not getSendData().empty(), "No send data configured. Use explicit scheme for one-way coupling.");` is invalid for `MultiCouplingScheme` (2bd54ee).

Reply:
dcd9a5b57e1605ba1fa7ef20f50b31171e8a2fa5 and 2df3ee2e8064f18852f5885ec3ff37600a733ac2 remove `initializeImplicit` from the leaf classes.

@uekerman I assume that the names of `checkForSend` and `checkAcceleration` could still be improved or we try to entirely get rid of the functions (see https://github.com/precice/precice/pull/689#discussion_r398735456).

Reply:
>  I assume that the names of checkForSend and checkAcceleration could still be improved

Yes, at least I don't understand what they mean without looking into the code :)

Reply:
A function with two or three lines that is only used once might not be worth a function :wink:

Reply:
I created a single function `checkConfiguration` to perform the checks. See 14bdedf0568724de48f20f6c1b1832ebb918b046. However, the clean solution would be to move this functionality out of the coupling scheme into the configuration.

Reply:
`checkConfiguration` is now called by `CouplingSchemeConfiguration`. See f90e40c.